frontend/src/components/shared/StreamlitMarkdown/StreamlitMarkdown.tsx

```diff
@@ -43,6 +43,60 @@ export interface Props {
   allowHTML: boolean
 }
 
+function createAnchorFromText(text: string | null): string {
+  const newAnchor = text
+    ?.toLowerCase()
+    .split(/[^A-Za-z0-9]/)
+    .filter(Boolean)
+    .join("-")
+  return newAnchor || ""
+}
+
+function createHeadingWithAnchor() {
+  let alreadyScrolled = false
+  return function({ tag, anchor, children }: any): ReactElement {
+    const ref = React.useCallback(
+      node => {
+        if (node !== null && !alreadyScrolled) {
+          const hash = window.location.hash.slice(1)
+          if (hash === (anchor || createAnchorFromText(node.textContent))) {
+            node.scrollIntoView(true)
+            alreadyScrolled = true
```

Comment:
Hm so this `alreadyScrolled` would prevent scrolling back to the same element when the anchor value changes. But if we somehow changed the hash and then changed it back, I don't believe this will get called / scroll into view. Alternatively if we changed the anchor and changed it back to match the current hash, it wouldn't scroll. Not sure how many scrolling use cases we want to get into with all this scrolling business. Could be something we refine as part of a future task but defer to @abhinand5. 

I'm not sure if this would also prevent us from scrolling to multiple elements? `alreadyScrolled` looks to be locally defined so each tag with an anchor would get it's own.

Reply:
Ok, I think I fixed this by just passing the anchor as the header's `id`. The dynamic scrolling handles scrolling on render, and after that, the browser handles everything else for us (clicking on a `#link`, manually changing the url, etc.)

> I'm not sure if this would also prevent us from scrolling to multiple elements? alreadyScrolled looks to be locally defined so each tag with an anchor would get it's own.

`createHeadingWithAnchor()` is only called once to initialize `HeaderWithAnchor`, so all instances of `<HeaderWithAnchor>` will use that variable stored in the closure. You're right though, it's confusing, I'll just make it global.